### PR TITLE
Add noreferences linter

### LINKED ISSUE #25: New Linter: references
* Fields should use Ref/Refs and not Reference/References
* For OpenShift, should not include Ref/Refs - a configuration option

### PR DESCRIPTION
**Pull request by krishagarwal278**
Added a new references linter to enforce that API field names use Ref/Refs suffixes instead of Reference/References.

What it does: 

- Flags fields ending with Reference and suggests changing to Ref
- Flags fields ending with References and suggests changing to Refs

Fixes #25

### CODE REVIEW

File: docs/linters.md [hunk 1]
```diff
@@ -555,6 +556,38 @@ If you prefer not to suggest fixes for pointers in required fields, you can chan
...
+By default, `references` is enabled and enforces this naming convention. The linter checks that:
+- Fields ending with 'Reference' are suggested to use 'Ref' instead
+- Fields ending with 'References' are suggested to use 'Refs' instead
+
+### Configuration
+
+```yaml
+lintersConfig:
+  references:
+    allowRefAndRefs: false | true # Allow Ref/Refs suffixes for OpenShift compatibility. Defaults to `false`.
```
Comment (JoelSpeed) [inline docs/linters.md]: We generally prefer enums over bools, so maybe lets change this to be more like `policy` config like we have in other places. Perhaps the values could be `AllowRefAndRefs` and `ForbidRefAndRefs`?

File: pkg/analysis/references/testdata/src/a/a.go [hunk 2]
```diff
@@ -0,0 +1,22 @@
+package a
+
+type TestSpec struct {
```
Comment (JoelSpeed) [inline pkg/analysis/references/testdata/src/a/a.go]: Can you try some field names where they contain the string `ref`/`refs` or `reference`/`references` but not at the end? Eg `Prefix` or `Prefereneces` perhaps could be in field names?

File: pkg/analysis/references/initializer.go [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+	)
+}
+
+func initAnalyzer(cfg *Config) (*analysis.Analyzer, error) {
+	return newAnalyzer(cfg), nil
+}
+
+// validateConfig validates the configuration for the references linter.
+func validateConfig(cfg *Config, fldPath *field.Path) field.ErrorList {
+	// No validation needed for this simple config
```
Comment (JoelSpeed) [inline pkg/analysis/references/initializer.go]: If we switch to an enum, I'd expect some validation here FWIW

**Review by JoelSpeed — COMMENTED** [review]
This is a great start, thanks for your contribution!

File: pkg/analysis/references/analyzer.go [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+	// If allowRefAndRefs is true (OpenShift), don't report errors for Ref/Refs fields
+	if !a.allowRefAndRefs {
+		// Check for fields ending with Ref or Refs (excluding those already handled above)
+		if fieldName != "Ref" && strings.HasSuffix(fieldName, "Ref") && !strings.HasSuffix(fieldName, "eRef") {
+			pass.Reportf(field.Pos(), "field %s should not use 'Ref' suffix", fieldName)
+		}
+
+		if fieldName != "Refs" && strings.HasSuffix(fieldName, "Refs") && !strings.HasSuffix(fieldName, "eRefs") {
+			pass.Reportf(field.Pos(), "field %s should not use 'Refs' suffix", fieldName)
+		}
```
Comment (sivchari) [inline pkg/analysis/references/analyzer.go]: What is `eRef` and `eRefs` ? why do we check the field with it ? Isn't it enough to only check `Ref` or `Refs` ?
I want some examples for test.

**Review by sivchari — COMMENTED** [review]
I want more test cases about references linter. And you mention OpenShift, but I think kube-api-linter should provide better practices for all developer who design Kubernetes API. Though original issue says OpenShift, I think it's better to publish this linter more general.

File: docs/linters.md (hunk 1 above)
Comment (everettraven) [reply docs/linters.md]: Nit: Let's also drop the OpenShift mention. This is a vendor-neutral project.

File: pkg/analysis/references/analyzer.go (hunk 4 above)
Comment (krishagarwal278) [reply pkg/analysis/references/analyzer.go]: replaced the initially added condition to check for letters/patterns before or after. Incorporated edge cases in test files 👍🏻

File: docs/linters.md [hunk 5]
```diff
@@ -555,6 +556,38 @@ If you prefer not to suggest fixes for pointers in required fields, you can chan
...
+```
+
+**Default behavior (allowRefAndRefs=false):**
+- Reports errors for fields ending with 'Reference' or 'References'
+- Also reports errors for fields ending with 'Ref' or 'Refs' (unless they were corrected from Reference/References)
+
+**OpenShift compatibility (allowRefAndRefs=true):**
```
[the remark this answers is not in the document]
Comment (krishagarwal278) [reply docs/linters.md]: Not unique to Openshift. Removed mention 👍🏻

File: .golangci.yaml [hunk 6]
```diff
@@ -31,6 +31,7 @@ linters:
     - govet
     - importas
     - ineffassign
+    - kubeapilinter
```
[the remark this answers is not in the document]
Comment (krishagarwal278) [reply .golangci.yaml]: Reverted that change 👍🏻  can't have kubeapilinter as a linter itself for this project 😆 

Was testing out/exploring different files in the linter and so might've committed this as a change too

File: docs/linters.md [hunk 7]
```diff
@@ -555,6 +556,41 @@ If you prefer not to suggest fixes for pointers in required fields, you can chan
 If you prefer not to suggest fixes for `omitempty` in required fields, you can change the `omitempty.policy` to `Warn` or `Ignore`.
 If you prefer not to suggest fixes for `omitzero` in required fields, you can change the `omitzero.policy` to `Warn` and also not to consider `omitzero` policy at all, it can be set to `Forbid`.
 
+## References
+
+The `references` linter ensures that field names use 'Ref'/'Refs' instead of 'Reference'/'References'.
+
+By default, `references` is enabled and operates in strict mode, prohibiting the use of 'Reference', 'References', 'Ref', or 'Refs' anywhere in field names
```
Comment (JoelSpeed) [inline docs/linters.md]: Default should be to allow Ref/Refs, but not Reference/References

File: pkg/analysis/namingconventions/analyzer.go [hunk 8]
```diff
@@ -36,9 +36,11 @@ type analyzer struct {
 	conventions []Convention
 }
 
-// newAnalyzer creates a new analysis.Analyzer for the namingconventions
+// NewAnalyzer creates a new analysis.Analyzer for the namingconventions
 // linter based on the provided config.
-func newAnalyzer(cfg *Config) *analysis.Analyzer {
+// This function is exported to allow other linters to wrap namingconventions
+// with fixed configuration.
+func NewAnalyzer(cfg *Config) *analysis.Analyzer {
```
Comment (JoelSpeed) [inline pkg/analysis/namingconventions/analyzer.go]: Are we certain this needs to be changed to exported? I thought we avoided that on the other PR

File: pkg/analysis/references/initializer_test.go [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+					Policy: references.PolicyForbidRefAndRefs,
+				},
+				expectedErr: "",
+			}),
+			Entry("With a valid references configuration with allowRefAndRefs=true", testCase{
+				config: &references.Config{
+					Policy: references.PolicyAllowRefAndRefs,
+				},
+				expectedErr: "",
+			}),
```
Comment (JoelSpeed) [inline pkg/analysis/references/initializer_test.go]: Case names are outdated here

File: pkg/analysis/references/doc.go [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+Similarly, fields ending with 'References' are suggested to use 'Refs' instead.
+
+Example configuration:
+
+**Default behavior (report errors for Reference/References):**
+```yaml
+lintersConfig:
+
+	references: {}
+
```
[the remark this answers is not in the document]
Comment (JoelSpeed) [reply pkg/analysis/references/doc.go]: Not yet actioned

File: pkg/analysis/references/analyzer.go [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+
+	// If policy is ForbidRefAndRefs, add conventions to forbid Ref/Refs anywhere
+	// Exclude patterns already handled by Reference/References above
+	if policy == PolicyForbidRefAndRefs {
+		conventions = append(conventions,
+			// Match "Refs" but not when part of "References"
+			namingconventions.Convention{
+				Name:             "forbid-refs",
+				ViolationMatcher: "Refs([^a-z]|$)",
+				Operation:        namingconventions.OperationInform,
```
Comment (JoelSpeed) [inline pkg/analysis/references/analyzer.go]: Should this be a Drop rather than inform?

File: docs/linters.md [hunk 12]
```diff
@@ -555,6 +556,41 @@ If you prefer not to suggest fixes for pointers in required fields, you can chan
...
+### Configuration
+
+```yaml
+lintersConfig:
+  references:
+    policy: ForbidRefAndRefs | AllowRefAndRefs # Defaults to `ForbidRefAndRefs`.
+```
+
+**Default behavior (policy: ForbidRefAndRefs):**
+- Reports errors for fields containing 'Reference' or 'References' anywhere and suggests replacing with 'Ref' or 'Refs'
```
[the remark this answers is not in the document]
Comment (krishagarwal278) [reply docs/linters.md]: Yikes, i guess I misunderstood the default policy to be `forbidRefAndRefs`. It's now `allowRefAndRefs` which makes sense because we want to allow `Ref` and `Refs` but not `Reference` and `References` as the default configuration option

File: pkg/analysis/references/analyzer.go (hunk 11 above)
Comment (krishagarwal278) [reply pkg/analysis/references/analyzer.go]: I think we should just inform here because user wants to `flag / forbidRefAndRefs` they may not necessarily want to drop that substring itself. 
For eg: We might not want `NodeRef` --> `Node`

dropping/swapping out ref(s) could be in scope for another story 👍🏻

File: pkg/analysis/references/testdata/src/a/a.go.golden [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+
+	ConfigRef string `json:"configRef"` // want `naming convention "reference-to-ref": field ConfigReference: field names should use 'Ref' instead of 'Reference'`
+
+	// Fields ending with References should be flagged  
+	NodeRefs []string `json:"nodeRefs"` // want `naming convention "references-to-refs": field NodeReferences: field names should use 'Refs' instead of 'References'`
+
+	ConfigRefs []string `json:"configRefs"` // want `naming convention "references-to-refs": field ConfigReferences: field names should use 'Refs' instead of 'References'`
+
+	// Fields with Reference at beginning should be flagged
+	RefCount int `json:"referenceCount"` // want `naming convention "reference-to-ref": field ReferenceCount: field names should use 'Ref' instead of 'Reference'`
```
[the remark this answers is not in the document]
Comment (krishagarwal278) [reply pkg/analysis/references/testdata/src/a/a.go.golden]: When testing locally, it's also fixing json. Latest commit has the new version of the file 👍🏻

File: pkg/analysis/references/initializer_test.go (hunk 9 above)
Comment (krishagarwal278) [reply pkg/analysis/references/initializer_test.go]: updated case names as per correct policy. The latest default policy is `AllowRefAndRefs`

File: pkg/analysis/references/analyzer.go [hunk 14]
```diff
@@ -0,0 +1,110 @@
...
+}
+
+// buildConventions creates the naming conventions based on the policy.
+func buildConventions(policy Policy) []namingconventions.Convention {
+	// Base convention: Replace "Reference" or "References" with "Ref" or "Refs"
+	// Using a single regex with optional 's' capture group to handle both cases
+	conventions := []namingconventions.Convention{
+		{
+			Name:             "reference-to-ref",
+			ViolationMatcher: "(?i)reference(s?)",
```
Comment (JoelSpeed) [inline pkg/analysis/references/analyzer.go]: Why do we make this case insensitive? Isn't that causing things like Preference to become Prefs, I don't think that's actually desired? I think we are better not having the case insensitivity aren't we?

File: pkg/analysis/references/analyzer.go [hunk 15]
```diff
@@ -0,0 +1,110 @@
...
+			Operation:        namingconventions.OperationReplacement,
+			Replacement:      "Ref$1",
+			Message:          "field names should use 'Ref' instead of 'Reference'",
+		},
+	}
+
+	// If policy is ForbidRefAndRefs, also flag Ref/Refs as problematic (no fix provided)
+	// This creates a two-step hint: fix Reference→Ref, but know that Ref also needs changing
+	if policy == PolicyForbidRefAndRefs {
+		conventions = append(conventions,
```
Comment (JoelSpeed) [inline pkg/analysis/references/analyzer.go]: I think this wants to replace, not append. That way we can avoid the intermediate steps.

Should be able to just have a single regex here I think

File: docs/linters.md [hunk 16]
```diff
@@ -560,25 +560,25 @@ If you prefer not to suggest fixes for `omitzero` in required fields, you can ch
...
-- **Also reports errors** for fields containing 'Ref' or 'Refs' anywhere (these should be dropped/removed entirely)
-- In this strict mode, the goal is to avoid all Ref/Refs/Reference/References in field names
-
-**Allow Ref/Refs (policy: AllowRefAndRefs):**
+**Default behavior (policy: AllowRefAndRefs):**
 - Reports errors for fields containing 'Reference' or 'References' and suggests replacing with 'Ref' or 'Refs'
 - **Allows** fields containing 'Ref' or 'Refs' without reporting errors
 
+**Strict mode (policy: ForbidRefAndRefs):**
+- Reports errors for fields containing 'Reference' or 'References' anywhere and suggests replacing with 'Ref' or 'Refs'
```
[the remark this answers is not in the document]
Comment (everettraven) [reply docs/linters.md]: +1, this should be dropping of the offending text.

File: pkg/analysis/references/analyzer.go (hunk 11 above)
Comment (everettraven) [reply pkg/analysis/references/analyzer.go]: Not sure I agree here. I think intention behind `forbidRefAndRefs` is actually to cause something like `NodeRef` -> `Node`.

If we think there is a reasonable use case to inform instead of suggest fixes, maybe we follow configuration precedence of other linters to allow a `Warn` and `SuggestFixes` configuration for this linter and we default to `SuggestFixes`?

File: docs/linters.md [hunk 17]
```diff
@@ -555,6 +556,41 @@ If you prefer not to suggest fixes for pointers in required fields, you can chan
 If you prefer not to suggest fixes for `omitempty` in required fields, you can change the `omitempty.policy` to `Warn` or `Ignore`.
 If you prefer not to suggest fixes for `omitzero` in required fields, you can change the `omitzero.policy` to `Warn` and also not to consider `omitzero` policy at all, it can be set to `Forbid`.
 
+## References
+
+The `references` linter ensures that field names use 'Ref'/'Refs' instead of 'Reference'/'References'.
```
[the remark this answers is not in the document]
Comment (JoelSpeed) [reply docs/linters.md]: Yeah I think that makes sense, `noreference` or `noreferences`

File: docs/linters.md (hunk 16 above)
[the remark this answers is not in the document]
Comment (krishagarwal278) [reply docs/linters.md]: @JoelSpeed @everettraven if y'all agree to dropping Ref/s too as offending text in ForbidRefAndRefs, can certainly make that change 👍🏻

File: pkg/analysis/references/analyzer.go (hunk 14 above)
Comment (everettraven) [reply pkg/analysis/references/analyzer.go]: @krishagarwal278 Could you verify that removing the case insensitivity results in successfully matching the json tags that look like `reference*` and `*Reference`?

File: pkg/analysis/references/analyzer.go [hunk 18]
```diff
@@ -0,0 +1,119 @@
...
+
+	case PolicyNoReferences:
+		// Drop any reference-related words from field names
+		// Using a regex that matches Ref/Refs/Reference/References at start or end
+		// At start: matches when followed by an uppercase letter, preserving that letter
+		// At end: matches at the end of the field name
+		return []namingconventions.Convention{
+			{
+				Name:             "no-references",
+				ViolationMatcher: "^([Rr]ef(?:erence)?s?)([A-Z])|([Rr]ef(?:erence)?s?)$",
```
Comment (JoelSpeed) [inline pkg/analysis/references/analyzer.go]:
```suggestion
				ViolationMatcher: "^([Rr]ef(?:erence)?s?)([A-Z])|(Ref(?:erence)?s?)$",
```
Shouldn't the second part of this not be case insensitive?

File: docs/linters.md (hunk 17 above)
[the remark this answers is not in the document]
Comment (JoelSpeed) [reply docs/linters.md]: Ok, @krishagarwal278 can you rename this to `noreferences` please

File: pkg/analysis/references/analyzer.go (hunk 18 above)
Comment (JoelSpeed) [reply pkg/analysis/references/analyzer.go]: I believe the first part is deliberate to pick up fields starting with Reference, cc @everettraven since I think he came up with this regex

File: pkg/analysis/noreferences/analyzer.go [hunk 19]
```diff
@@ -0,0 +1,113 @@
...
+				ViolationMatcher: "^[Rr]eference(s?)|Reference(s?)$",
+				Operation:        namingconventions.OperationReplacement,
+				Replacement:      "Ref$1$2",
+				Message:          "field names should use 'Ref' instead of 'Reference'",
+			},
+		}
+
+	case PolicyNoReferences:
+		// Drop any reference-related words from field names
+		// At start: matches [Rr]ef(erence)?s? followed by uppercase letter, preserving that letter ($2)
```
Comment (JoelSpeed) [inline pkg/analysis/noreferences/analyzer.go]:
```suggestion
		// At start: matches [Rr]ef(erence)?s? followed by uppercase letter, preserving that letter ($3)
```

File: pkg/analysis/noreferences/doc.go [hunk 20]
```diff
@@ -0,0 +1,43 @@
...
+	  policy: PreferAbbreviatedReference
+
+```
+**Strict mode (forbid Ref/Refs in field names):**
+```yaml
+lintersConfig:
+
+	noreferences:
+	  policy: NoReferences
+
```
Comment (JoelSpeed) [inline pkg/analysis/noreferences/doc.go]:
```suggestion
	noreferences:
	  policy: NoReferences
```

File: golangci-lint-kube-api-linter
Comment (JoelSpeed) [inline golangci-lint-kube-api-linter]: This should not be committed

File: docs/linters.md [hunk 21]
```diff
@@ -555,6 +556,53 @@ If you prefer not to suggest fixes for pointers in required fields, you can chan
...
+  - `PreferenceType` → no change (middle - not flagged)
+
+**NoReferences mode:**
+- Removes all reference-related words (Ref/Refs/Reference/References) from the beginning or end of field names
+- Like PreferAbbreviatedReference, this only matches at word boundaries to avoid false positives
+- Examples:
+  - `RefNode` → `Node` (start)
+  - `NodeRef` → `Node` (end)
+  - `ReferenceName` → `Name` (start)
+  - `ConfigReferences` → `Config` (end)
```
Comment (JoelSpeed) [inline docs/linters.md]: Looking at the test, I believe the `s` is preserved isn't it?
